### PR TITLE
Central Pumping and water gates

### PR DESCRIPTION
**Pull request by Torca2001**
#44 Implemented a functional central pumping system that requires water gates to function.

Transitioned existing Mechtrauma pumps to use the new water gate system that prevents cheesing of the water connection

Water gates that contain items will cause them to be blocked and prevent them from functioning.

**Changes were made to some pump connections and gates, existing subs will have to be checked**

Also appears that #61 is fixed with this

### CODE REVIEW

**Comment by Torca2001** [comment]
Also fixed pumps set_state pins fighting with the xml causing the pump to constantly turn on or off.
Fixed the backup batteries in pumps to only draw power when the grid fails to meet the min voltage requirement
Modified the pump gui to indicate it is using the battery by rapidly flashing the power indicator. 
(Possibly change the power colour but that requires creating a new sprite for it)

**Comment by TBN-MapleWheels** [comment]
Merged. Requires some cleanup that's easier to do post-merge than pre-merge. Note: Does not fix #61 fundamentally as it's not a flaw with the pumping system component but the Reflection lookup in Luatrama that requires a custom solution using Modding TK..

File: SharedProject/SharedSource/powerConnections.cs [hunk 1]
```diff
@@ -22,6 +22,7 @@ public partial class Plugin : IAssemblyPlugin
             "steam",
```
Comment (TBN-MapleWheels) [inline SharedProject/SharedSource/powerConnections.cs:22]: This entire file needs to be merged into the Plugin.cs file under the Shared Project.

File: SharedProject/SharedSource/WaterGate.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    partial class WaterGate : Powered {
+
+        public float MaxFlow
+        {
+            get => maxFlow;
+            set => maxFlow = value;
+        }
+        private float maxFlow = 100.0f;
+
+        [Serialize(false, IsPropertySaveable.Yes, alwaysUseInstanceValues: true)]
```
Comment (TBN-MapleWheels) [inline SharedProject/SharedSource/WaterGate.cs:26]: There's no reason not to just use auto-properties here.

File: SharedProject/SharedSource/CentralPump.cs [hunk 3]
```diff
@@ -8,17 +8,29 @@
...
+using System.Globalization;
 
-namespace Barotrauma.Items.Components 
+namespace Mechtrauma
 {
-    class CentralPump : Pump {
+    class CentralPump : BatteryPump {
+
+        [Editable, Serialize(false, IsPropertySaveable.No, description: "Should an external water gate be required to function", alwaysUseInstanceValues: true)]
+        public bool NeedWaterGate
```
Comment (TBN-MapleWheels) [inline SharedProject/SharedSource/CentralPump.cs:18]: Use auto-properties in this class.

File: SharedProject/SharedSource/CentralPump.cs [hunk 4]
```diff
@@ -28,26 +40,62 @@ public float HullPercentage
...
             }
 
             if (TargetLevel != null) {
-                flowPercentage = ((float)TargetLevel - HullPercentage) * 10.0f;
+                flowPercentage = ((float)TargetLevel - HullPercentage);
             }
 
-            if (!HasPower) {
-                return;
+            Item? battery = getBackupBattery();
```
Comment (TBN-MapleWheels) [inline SharedProject/SharedSource/CentralPump.cs:50]: Ensure all methods, types and files use Pascal Case.

**Review by TBN-MapleWheels — COMMENTED** [review]
Minor issues to fix, some consolidation required.

**Pull request merged**